Framework: Substitute crypto polyfill dependants with smaller stand-ins

aduth opened the pull request:
A few of our dependencies, namely `component-uid` and `js-sha1`, themselves have dependencies on `crypto`:

https://github.com/emn178/js-sha1/blob/c724643/src/sha1.js#L52-L53
https://github.com/component/uid/blob/2d912b8/index.js#L45 (It appears they try to avoid the polyfill, but it _is_ included in a Webpack build)

`crypto` is a [Node built-in module](https://nodejs.org/api/crypto.html) which is polyfilled in the browser by [crypto-browserify](https://github.com/crypto-browserify/crypto-browserify). This polyfill is quite substantial in size.

The changes here seek to find suitable replacements for each: `uuid` and [`hash.js`](https://www.npmjs.com/package/hash.js) respectively, which are very reasonable in size.

_Aside: I first suspected `uuid` to also [require `crypto`](https://github.com/kelektiv/node-uuid/blob/72fbabbf00192817bdce7073e323361f5e7bdd67/lib/rng.js#L4), but it turns out they bundle a [browser-specific variant](https://github.com/kelektiv/node-uuid/blob/72fbabbf00192817bdce7073e323361f5e7bdd67/lib/rng-browser.js) without the dependency_.

The difference is pretty noticeable: 472kb smaller minified (112kb gzipped) in the main `build.js`.

**Before:**

```
wp-calypso (master) $ wc -c public/build.b1e9a12a27f881b1406e.js
 7190661 public/build.b1e9a12a27f881b1406e.js
wp-calypso (master) $ gzip-size public/build.b1e9a12a27f881b1406e.js --raw
1275631
```

**After:**

```
wp-calypso (master) $ wc -c public/build.10d51aa5c5ffef307acf.js
 6718680 public/build.10d51aa5c5ffef307acf.js
wp-calypso (master) $ gzip-size public/build.10d51aa5c5ffef307acf.js --raw
1163240
```

__Testing instructions:__

Verify against regressions for the two modules affected:

- Sync handler (used for caching response data on the posts list)
- Popover (tooltips, ellipsis menu, etc)

Ensure unit tests pass:

```
npm test
```

gwwar commented:
I'm not sure if it's related but devdocs/blocks isn't loading for me on this branch.

aidvu commented:
@gwwar don't think so. There are 2-3 components in the blocks that are broken I believe. One is fixed https://github.com/Automattic/wp-calypso/commit/925af9200a9679921d052d291b4ec5ba70488bbc. Can take a look on Monday.

retrofox commented on client/components/popover/index.jsx:
```diff
@@ -361,7 +361,7 @@ class Popover extends Component {
 
 	setPopoverId( id ) {
-		this.id = id || `pop__${ uid( 16 ) }`;
+		this.id = id || `pop__${ uuid() }`;
```
This line is almost unneeded. If we think that it's better to get rid of it let's do it. I could do it in a next PR as well.

DanReyLop replied:
Yeah, `lodash/uniqueId` seems better (and simpler) for this case. It actually guarantees that it will be unique, while `uuid` can't.

retrofox commented on client/lib/wp/sync-handler/utils.js:
```diff
@@ -1,11 +1,10 @@
  * External dependencies
  */
-import sha1 from 'js-sha1';
+import sha1 from 'hash.js/lib/hash/sha/1';
```
in fact, we could get rid of sync-handler as well. In this way, we could clean this module completely.

aduth replied:
>in fact, we could get rid of sync-handler as well. In this way, we could clean this module completely.

Yeah, I figured sync handler is on its way out. Most of the pieces are already there with the newer custom post type screen (i.e. caching post responses for future use). I think the remaining issue is whether we need to preserve existing UIs (especially with newer features like re-Publicize) and how we extend CPT components to support those cases. Regardless, it's a bigger issue than one I want to address here, and I think it'll be good to keep the sync handler around until a full replacement can be made.

aduth commented on client/components/popover/index.jsx:
```diff
@@ -361,7 +361,7 @@ class Popover extends Component {
 
 	setPopoverId( id ) {
-		this.id = id || `pop__${ uid( 16 ) }`;
+		this.id = id || `pop__${ uuid() }`;
```
A static incrementing number could work just as well, yeah. (Or nothing at all)

blowery commented:
> The difference is pretty noticeable: 472kb smaller minified (112kb gzipped)

😲

alisterscott commented:
I did some testing of this and couldn't find any regressions to areas mentioned 👍

aduth commented:
Rebased and replaced the Popover's `uuid` with Lodash `uniqueId`, regenerated shrinkwrap once more, did a final round of testing, and will merge shortly.

gwwar commented:
@aduth @tyxla heads up on overlapping changes in https://github.com/Automattic/wp-calypso/pull/17560

aduth commented:
Thanks for catching the overlap @gwwar

tyxla commented:
Thanks @aduth @gwwar!
